### PR TITLE
ASR: assign dimensions of ttype of ArrayReshape node correctly

### PR DESCRIPTION
**Pull request by gxyd**
helps fix runtime error in `integration_tests/intrinsics_159.f90`

[EDIT]: the same change works for `integration_tests/parameter_11.f90` as well.

### Minimal example to understand the change

For the below Fortran program
```f90
program main
    integer(8) :: x_dp(2, 3)
    integer :: i
    x_dp = reshape([ (i, i = 1, 6) ], [2, 3]) ! simplifier_pass branch outputs 1, this branch corrects it to 2
    print *, x_dp(2, 1)
end program main
```

the output currently produced in simplifier_pass branch is `1`, which is I think incorrect. This branch fixes that output.

### ASR changes

ASR of the above fortran produced in current main (and hence in simplifier_pass branch is):
```clj
(TranslationUnit
    (SymbolTable
        1
        {
            main:
                (Program
                    (SymbolTable
                        2
                        {
                            i:
                                (Variable
                                    2
                                    i
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4)
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            x_dp:
                                (Variable
                                    2
                                    x_dp
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Array
                                        (Integer 8)
                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                        (IntegerConstant 2 (Integer 4) Decimal))
                                        ((IntegerConstant 1 (Integer 4) Decimal)
                                        (IntegerConstant 3 (Integer 4) Decimal))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    main
                    []
                    [(Assignment
                        (Var 2 x_dp)
                        (Cast
                            (ArrayReshape
                                (ArrayConstructor
                                    [(ImpliedDoLoop
                                        [(Var 2 i)]
                                        (Var 2 i)
                                        (IntegerConstant 1 (Integer 4) Decimal)
                                        (IntegerConstant 6 (Integer 4) Decimal)
                                        ()
                                        (Integer 4)
                                        (ArrayConstant
                                            24
                                            [1, 2, 3, 4, 5, 6]
                                            (Array
                                                (Integer 4)
                                                [((IntegerConstant 1 (Integer 4) Decimal)
                                                (IntegerConstant 6 (Integer 4) Decimal))]
                                                FixedSizeArray
                                            )
                                            ColMajor
                                        )
                                    )]
                                    (Array
                                        (Integer 4)
                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                        ())]
                                        DescriptorArray
                                    )
                                    (ArrayConstant
                                        24
                                        [1, 2, 3, 4, 5, 6]
                                        (Array
                                            (Integer 4)
                                            [((IntegerConstant 1 (Integer 4) Decimal)
                                            (IntegerConstant 6 (Integer 4) Decimal))]
                                            DescriptorArray
                                        )
                                        ColMajor
                                    )
                                    ColMajor
                                )
                                (ArrayPhysicalCast
                                    (ArrayConstant
                                        8
                                        [2, 3]
                                        (Array
                                            (Integer 4)
                                            [((IntegerConstant 1 (Integer 4) Decimal)
                                            (IntegerConstant 2 (Integer 4) Decimal))]
                                            FixedSizeArray
                                        )
                                        ColMajor
                                    )
                                    FixedSizeArray
                                    DescriptorArray
                                    (Array
                                        (Integer 4)
                                        [((IntegerConstant 1 (Integer 4) Decimal)
                                        (IntegerConstant 2 (Integer 4) Decimal))]
                                        DescriptorArray
                                    )
                                    ()
                                )
                                (Array
                                    (Integer 4)
                                    [(()
                                    ())]
                                    DescriptorArray
                                )
                                ()
                            )
                            IntegerToInteger
                            (Array
                                (Integer 8)
                                [((IntegerConstant 1 (Integer 4) Decimal)
                                (IntegerConstant 2 (Integer 4) Decimal))
                                ((IntegerConstant 1 (Integer 4) Decimal)
                                (IntegerConstant 3 (Integer 4) Decimal))]
                                FixedSizeArray
                            )
                            ()
                        )
                        ()
                    )
                    (Print
                        [(StringFormat
                            ()
                            [(ArrayItem
                                (Var 2 x_dp)
                                [(()
                                (IntegerConstant 2 (Integer 4) Decimal)
                                ())
                                (()
                                (IntegerConstant 1 (Integer 4) Decimal)
                                ())]
                                (Integer 8)
                                ColMajor
                                ()
                            )]
                            FormatFortran
                            (Character -1 0 ())
                            ()
                        )]
                        ()
                        ()
                    )]
                )
        })
    []
)
```

This PR correct's the dimensions of `ArrayReshape` node. i.e. produces a diff in ASR as:
```diff
119a120,121
>                                     ())
>                                     (()
```

### CODE REVIEW

**Comment by gxyd** [comment]
There might be a question of whether this PR (or atleast some changes of this PR) should be against main OR simplifier_pass branch. As there is use of `create_array_type_with_empty_dims` as well, which might not make it easy to create this PR against main.

For me, I'm more happy with this PR against simplifier_pass.

**Comment by gxyd** [comment]
Interesting, the test i.e. `intrinsics_159.f90` passes on macOS (ran locally "Check Debug build" CI job), while it fails on Linux.

**Comment by gxyd** [comment]
I've uncommented `parameter_11` as well, and that also passes now in macOS atleast.

**Comment by gxyd** [comment]
I re-ran the job and still the above two tests fail on Linux, i.e. `intrinsics_159` and `parameter_11`.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -5215,40 +5215,42 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
-         }
-         newshape = ASRUtils::cast_to_descriptor(al, newshape);
-         return ASR::make_ArrayReshape_t(al, x.base.base.loc, array, newshape, empty_type, nullptr);
+                                    ASRUtils::type_get_past_pointer(ASRUtils::expr_type(array))),
+                                    &dims, array_physical_type, true);
+        } else {
+            ASR::Array_t* arr_type = ASR::down_cast<ASR::Array_t>(ASRUtils::expr_type(newshape));
+            size_t n_dims = ASR::down_cast<ASR::IntegerConstant_t>(arr_type->m_dims[0].m_length)->m_n;
+            empty_type = ASRUtils::create_array_type_with_empty_dims(al, n_dims,
+                ASRUtils::type_get_past_allocatable(ASRUtils::type_get_past_pointer(ASRUtils::expr_type(array))));
```
Comment (kmr-srbh) [inline src/lfortran/semantics/ast_common_visitor.h:5250]: Could you please explain why creating an array with empty type here fixes the issue? I think the previous code did that (with `dims`)? :)

Comment (gxyd) [reply src/lfortran/semantics/ast_common_visitor.h:5250]: If you see the change made in the PR description with ASR, you'll notice that the dimensions of the output "Reshape", i.e. it's ttype is incorrect currently. It gives Reshape a ttype with dimension of always 1. This PR intends to fix that.

File: src/libasr/asr_utils.cpp [hunk 2]
```diff
@@ -1649,6 +1649,21 @@ void make_ArrayBroadcast_t_util(Allocator& al, const Location& loc,
...
+        Location loc; loc.first = 1, loc.last = 1;
+        empty_dim.loc = loc;
+        empty_dim.m_length = nullptr;
+        empty_dim.m_start = nullptr;
+        empty_dims.push_back(al, empty_dim);
+    }
+    return ASRUtils::make_Array_t_util(al, value_type->base.loc,
+        ASRUtils::extract_type(value_type), empty_dims.p, empty_dims.size());
+}
+
```
[the remark this answers is not in the document]
Comment (gxyd) [reply src/libasr/asr_utils.cpp:1666]: I've tried making this (partial) change to use `duplicate_type_with_empty_dims`, let's see if the tests pass on Linux or not in CI.

**Comment by kmr-srbh** [comment]
The example in the PR description prints the correct output `2` with the changes, but the test `intrinsics_159.f90` fails. I think https://github.com/lfortran/lfortran/issues/4738#issuecomment-2328294048 might be the MRE for it.

**Comment by gxyd** [comment]
> The example in the PR description prints the correct output `2` with the changes, but the test `intrinsics_159.f90` fails. I think [#4738 (comment)](https://github.com/lfortran/lfortran/issues/4738#issuecomment-2328294048) might be the MRE for it.

Seems like that MRE also succeeds on macOS, but causes problem only on Linux.

**Comment by kmr-srbh** [comment]
I think the memory related issues are behaving differently on macOS and Linux. :thinking:

**Pull request closed without merging**